File: longestWordinDict.py
from collections import deque
class TrieNode(object):
    def __init__(self):
        self.word = None
        self.branches = [None] * 26


class Solution(object):
    # Time Complexity : O(nl) where n is the number of words and l is the average length of the words
    # Space Complexity : O(nl), where n is the number of words in the trie and l is the average length of the words
    # Did this code successfully run on Leetcode : Yes
    # Any problem you faced while coding this : No

    # Your code here along with comments explaining your approach

    # The given words are first inserted in the trie, then searched.
    # A BFS is performed to search all nodes level by level while also maintaining
    # smaller lexicographical order, In the end the longest word is returned.
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        curr = self.root
        for i in word:
            temp = ord(i) - ord('a')
            if curr.branches[temp] == None:
                curr.branches[temp] = TrieNode()
            curr = curr.branches[temp]
        curr.word = word

    def longestWord(self, words):
        # adding to the trie
        for i in words:
            self.insert(i)
        # queue init
        q = deque()
        q.appendleft(self.root)
        curr = None
        # BFS with small lexicographical order
        while len(q) != 0:
            curr = q.pop()
            for j in range(25, -1, -1):
                level = curr.branches[j]
                if level != None and level.word != None:
                    q.appendleft(level)
        return curr.word

File: longestWordinDict.java
// Time Complexity : O(N sub(l)), l is the length of each word in the trie
// Space Complexity : O(N sub(l)), l is the length of each word in the trie
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

public class longestWordinDict {
    class Trie {
        class Node {
            char ch;
            Node[] children = new Node[26];
            boolean isWord = false;
            
            public Node (char c) {
                ch = c;
                isWord = false;
            }
        }
        
        Node root;
        int len;
        String maxWord;
        
        public Trie() {
            root = new Node('\0');
            len = 0;
            maxWord = "";
        }
        
        public void insert(String s) {
            Node curr = root; 
            boolean flag = true;
            
            for(int i=0; i<s.length(); i++){
                char c = s.charAt(i);
                if (curr.children[c-'a']!=null) {
                    if(!curr.children[c-'a'].isWord) {
                        flag = false;
                    }
                    curr = curr.children[c-'a'];
                } else {
                    if(i!=s.length()-1) {
                        flag = false;
                    }
                    curr.children[c-'a'] = new Node(c);
                    curr = curr.children[c-'a'];
                }
            }
            curr.isWord = true;
            
            if (flag) {
                if (s.length()>len) {
                    len = s.length();
                    maxWord = s;
                } else if (s.length()==len) {
                    if(s.compareTo(maxWord) < 0)
                        maxWord = s;
                }
            }
        }
    }
    
    public String longestWord(String[] words) {

        Trie trie = new Trie();
        Arrays.sort(words);
        for(String s: words){
            trie.insert(s);
        }
        return trie.maxWord;
    }
}
